### PR TITLE
Refactor address state validation

### PR DESCRIPTION
**Pull request by cedum**
**Description**

Deprecates and refactors state validation by extracting out the normalization and validation logic into a separate service-like object, which might be also user-provided, for example:
```
Spree::Address.state_validator_class = UserProvided::Address::StateValidator
```

The following private methods have been deprecated from `Spree::Address` model:
- `state_validate`
- `validate_state_matches_country`

This PR introduces the following configuration flag `Spree::Config.use_legacy_address_state_validator` that enables to switch from the old state validation behavior to the new in a smooth way.
For newly generated applications, this flag is set to `false`, which means it will use the new state validator class.
For existing applications upgrading to Solidus v2.11 it is set to `true`, hence keeping the same behavior as in the previous Solidus versions.

**Checklist:**
- [ ] Write a recap in this PR description with the normalization and validation rules
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [x] I have added a detailed description into each commit message
- [ ] I have updated Guides and README accordingly to this change (if needed)
- [x] I have added tests to cover this change (if needed)

### CODE REVIEW

**Comment by cedum** [comment]
Moved the PR back to WIP 🚧. It requires some more work. I'll try to make it less break. Besides this, by digging deeper into it I see I misinterpreted some state behaviours and also introduced a few bugs 😅

File: core/app/models/spree/address.rb
Comment (kennyadsl) [inline core/app/models/spree/address.rb]: Normalization is also happening into the validation, right? Why it is needed to perform it here as well?

File: core/app/models/spree/address/state_validator.rb
Comment (kennyadsl) [inline core/app/models/spree/address/state_validator.rb]: What about passing this class to the `initialize` of this class?

```ruby
def initialize(address, address_normalizer_class = Spree::Address.state_normalizer_class)

# and in the caller
self.class.state_validator_class.new(self, self.class.state_normalizer_class).perform!
```

This way it will still be possible to change the class as an address level preference but we can also have more flexibility if needed for some customization or testing needs.

File: core/app/models/spree/address/state_normalizer.rb
Comment (kennyadsl) [inline core/app/models/spree/address/state_normalizer.rb]: Personal preference, but I'd use `def call` here. Did you have a specific reason to use `perform!` instead?

File: core/app/models/spree/address/state_validator.rb
Comment (cedum) [reply core/app/models/spree/address/state_validator.rb]: This wasn't intended to be here. I have moved it out into `validate` block in `Spree::Address`. Forgot to remove it from here :|

File: core/app/models/spree/address/state_normalizer.rb
Comment (cedum) [reply core/app/models/spree/address/state_normalizer.rb]: I have followed the convention of other service-like objects in Solidus that are using `perform` as their main action method. I have no preferences though.
Probably it's worth defining a project-wise convention for this kind of objects

File: core/app/models/spree/address.rb
Comment (cedum) [reply core/app/models/spree/address.rb]: The original idea was to move it here from the validator service object so they can be completely decoupled. Just forgot to remove it from there (see also the comment bellow related to this one)

**Comment by cedum** [comment]
Ready for a second review 🙏 

@kennyadsl I've researched a few methods to deprecate them by showing a warning message only if they have been monkey patched, but the final solution was quite hacky and not working in all cases since there're multiple ways to patch them.
Being both of them private methods, you think it's worth the effort to deprecate them before removing? I've invested some time into it but I wasn't able to find a +/- clean way to deprecate them.

**Comment by cedum** [comment]
Rebased against master and resolved conflicts

**Comment by aldesantis** [comment]
@cedum can you provide some info about the current state of the deprecation warning in this PR? It looks like everything should work properly if the private methods are being overridden, am I mistaken?

File: core/app/models/spree/address.rb
Comment (aldesantis) [inline core/app/models/spree/address.rb:29]: What if we called the custom methods instead of the new behavior in case the methods have been added by the user in a decorator? We may even do that in the `state_normalizer_class` and `state_validator_class` themselves.

**Comment by kennyadsl** [comment]
I recall we decided to merge this in 3.0 but I think it wasn't a good decision since, if possible, we should make this optional and deprecate the old behavior in 2.11. This way people have time to fix this in their application before the 3.0 upgrade.

File: core/app/models/spree/address.rb
Comment (kennyadsl) [inline core/app/models/spree/address.rb:231]: Until the preference is not changed users will see 3 deprecation warnings. It would be awesome to find a way to only print one deprecation warning in that case, but keep the individual ones if these methods are called from somewhere else. I'm not sure it's a good idea but what about passing an arg to these deprecated methods to conditionally silence the deprecation warning?

This way we can do something like:

```
      if Spree::Config.legacy_address_state_validator
        state_validate(silence_deprecation: true)
        validate_state_matches_country(silence_deprecation: true)
      else
```

Users still see the main deprecation warning if using default code and the specific ones if they used `state_validate` or `validate_state_matches_country` in other ways. I'm not particularly concerned about changing the arity of those methods since they'll go away soon. 

What's your thoughts?

File: core/lib/spree/testing_support/dummy_app.rb
Comment (kennyadsl) [inline core/lib/spree/testing_support/dummy_app.rb:122]: Can we alphabetize this new preference?

File: core/app/models/spree/address.rb
Comment (cedum) [reply core/app/models/spree/address.rb:231]: 👍 agree!
By introducing a new parameter it will break any existing prepended patches of these methods. I'm not aware of a way to introduce a new parameter w/o being breaking. I'm looking if there's another way

Comment (kennyadsl) [reply core/app/models/spree/address.rb:231]: Maybe just setting an accessor on the class?

Comment (cedum) [reply core/app/models/spree/address.rb:231]: Another way might be to wrap the two calls in a `silence` block, this avoids polluting the Address class API with new attributes/methods. The downside is it will silence any eventual deprecations generated by these two calls.
What do you think?

```ruby
validate do
  if Spree::Config.legacy_address_state_validator
    Spree::Deprecation.silence do
      state_validate
      validate_state_matches_country
    end
  else
    self.class.state_validator_class.new(self).perform
  end
end
```

Comment (kennyadsl) [reply core/app/models/spree/address.rb:231]: It's more elegant for sure but I think (might be worth a double-check) we are not using `Spree::Deprecation.silence` anywhere in the codebase (outside specs). I'm not completely sure why to be honest and if this could cause issues that we can't anticipate at the moment. Anyway, for now this seems like the best solution let's go head with this and let's see what the other core team member think about it.

Comment (cedum) [reply core/app/models/spree/address.rb:233]: just discovered that `silence` seems to not be thread-safe 🤦 Here's a simple test case that sometimes leaves silenced class attribute set to `true`: https://gist.github.com/cedum/8315ebdd59781ee7fcc7f452a405542f
I'm refactoring this to a simple instance variable switch 😄, something like this:
```ruby
validate do
  if Spree::Config.legacy_address_state_validator
    begin
      @silence_state_deprecations = true
      state_validate
      validate_state_matches_country
    ensure
      @silence_state_deprecations = false
    end
  else
    self.class.state_validator_class.new(self).perform
  end
end
```

**Review by kennyadsl — APPROVED** [review]
👍

**Comment by kennyadsl** [comment]
@cedum we discussed this in the core meeting and we agreed that it would be much better if we have the fix for #2417 and #3098 in a different commit, just for documentation in case someone cannot update and wants to just understand what changed to backport that in their own application. Do you think it's doable?

**Comment by cedum** [comment]
> @cedum we discussed this in the core meeting and we agreed that it would be much better if we have the fix for #2417 and #3098 in a different commit, just for documentation in case someone cannot update and wants to just understand what changed to backport that in their own application. Do you think it's doable?

Yep, makes sense! I'm on it

File: core/app/models/spree/address.rb
Comment (cedum) [reply core/app/models/spree/address.rb:287]: ok boss

**Comment by cedum** [comment]
I've extracted the fix for #3098 into a separate PR (merged into master), this way it should be easier to backport the fix on stores unable to upgrade. 
I've just rebased also this PR and I think it's ready for another review 🙏

**Review by kennyadsl — APPROVED** [review]
👍 Thanks @cedum !

File: core/lib/spree/core/engine.rb
Comment (aldesantis) [inline core/lib/spree/core/engine.rb:82]: Can we improve this message to explain that validation will be done through an external class moving forward?

**Review by aldesantis — CHANGES_REQUESTED** [review]
@cedum just left a comment on a deprecation message, but other than that it's good to go!

File: core/lib/spree/core/engine.rb
Comment (cedum) [reply core/lib/spree/core/engine.rb:82]: @aldesantis what do you think about this message?

> DEPRECATION WARNING: Spree::Config.use_legacy_address_state_validator set to true has been deprecated and will be removed in Solidus 3.0. The Spree::Address state validation has been extracted into a configurable external class. Switch Spree::Config.use_legacy_address_state_validator to true to start using the external validator class.

Comment (aldesantis) [reply core/lib/spree/core/engine.rb:82]: @cedum sounds great!

**Review by spaghetticode — APPROVED** [review]
@cedum thank you 👍

**Review by aldesantis — APPROVED** [review]

**Pull request merged**
